backend/core/src/main/java/org/sonarsource/sonarlint/core/file/ServerFilePathsProvider.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * SonarLint Core - Implementation
+ * Copyright (C) 2016-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.file;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CancellationException;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonarsource.sonarlint.core.ServerApiProvider;
+import org.sonarsource.sonarlint.core.commons.Binding;
+import org.sonarsource.sonarlint.core.commons.progress.SonarLintCancelMonitor;
+import org.sonarsource.sonarlint.core.serverapi.ServerApi;
+
+import static java.util.stream.Collectors.toList;
+
+@Named
+@Singleton
+public class ServerFilePathsProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(ServerFilePathsProvider.class);
+
+  private final ServerApiProvider serverApiProvider;
+  private final Map<Binding, Path> cachedResponseFilePathByBinding = new HashMap<>();
+  private final Path cacheDirectoryPath;
+  private final Cache<Binding, List<Path>> temporaryInMemoryFilePathCacheByBinding;
+
+  public ServerFilePathsProvider(ServerApiProvider serverApiProvider, Path storageRoot) {
+    this.serverApiProvider = serverApiProvider;
+    this.cacheDirectoryPath = storageRoot.resolve("cache");
+    this.temporaryInMemoryFilePathCacheByBinding = CacheBuilder.newBuilder()
+      .expireAfterWrite(Duration.of(1, ChronoUnit.MINUTES))
+      .maximumSize(3)
```

Comment:
These two configurations might be calibrated

Reply:
I gave small values so as not to tire the memory too much. We can even think of smaller values.